### ISSUE TITLE
[request] Capturing from games

### ISSUE THREAD

**Issue by Shavitush** [ShareX/ShareX#848]
I really like the software and I think that we also need some love for gamers.  
How about making capturing from fullscreen games possible? It used to work at Windows 7 with Aero disabled, but now with the new Windows release, it's kinda over. A black screen is being seen when capturing from a fullscreen app, which is mainly games.

**Comment by HitomiTenshi** [comment]
I don't really think ShareX is designed for a gaming purpose, though recent Steam Greenlight may indicate otherwise. I would suggest going into your `[Game]`'s video settings and enabling `Borderless Windowed` / `Windowed Fullscreen`, which will fix your black-screen issue.

**Comment by Shavitush** [comment]
Obviously, but in fast paced games such as CS:GO, osu! etc you can't play borderless with WIndows' DWM enabled, it adds some massive unbearable input lag.

**Comment by HitomiTenshi** [comment]
I have to admit that I have experienced lag in Windows 7 Borderless before, but I tried playing CS:GO on Windows 10 with `Borderless` + `Raw Mouse Input` yesterday and it had absolutely no input lag at all. It was actually faster than `Fullscreen` + `No Vsync` + `Raw Mouse Input`.

You should try `Borderless Fullscreen` again with Windows 10. I assume this dramatic change has something to do with Windows 10 not having Aero (which was managing Vsync in Windows 7).

**Comment by Jaex** [comment]
I'm gamer too and I hope someone will add DirectX/OpenGL support :(

**Comment by Shavitush** [comment]
Argh that's a bummer. Guess we have to wait for some champion to send a pull request ;D

**Comment by Jaex** [comment]
It not gonna happen this is extremely difficult thing to make. Unless somehow Windows 10 have api that allows to use their Xbox screen capture feature.

**Comment by master131** [comment]
@Jaex You may want to consider implementing bits and piece's of spazzarama's Direct3DHook project which is a .NET DirectX 9/10/11 hooking library. It contains a screenshot capture sample project too. Might be a bit of work adapting to ShareX but it is a possibility.

https://github.com/spazzarama/Direct3DHook

**Comment by Jaex** [comment]
I tried it before many times and it was not working in majority of games and was crashing games and other such problems. And that project not actively worked on.

**Comment by master131** [comment]
Hmm, that's unfortunate then. What about the Desktop Duplication API?

**Comment by Jaex** [comment]
I don't know is it allow to capture from DirectX games. But I don't think so. Just because it using DirectX to do stuff not means it can capture DirectX games too.

**Comment by master131** [comment]
Perhaps if I ever get the time I might fork ShareX (or Direct3DHook) and try get a working DirectX hook based off Open Broadcasting Software's Graphics Capture plugin which is proven to work and be pretty reliable (unless someone is willing to do it).

https://github.com/jp9000/OBS/tree/master/GraphicsCapture

**Comment by Jaex** [comment]
https://github.com/jasonpang/desktop-duplication-net

I found this but because I don't have much knowledge about DirectX programming, I can't figure out is it can do what we need.

**Comment by master131** [comment]
I already tried using the Desktop Duplication API from the following sample:
https://github.com/sharpdx/SharpDX-Samples/tree/master/Desktop/Direct3D11.1/ScreenCapture

If you look at my previous post, you can see a screenshot I took using the sample which isn't rendered correctly so it's no use.

**Comment by Jaex** [comment]
Which screenshot?

**Comment by 3941** [comment]
While I'd love to see built-in support for this, you can easily help yourself by using another software like MSI Afterburner to make screenshots of games.
Then use ShareX's watchfolder feature to monitor the screenshot folder of your third party software and upload automatically.

Unless ShareX becomes really, really big, that's also your best bet on not being caught for "cheat injection" or the like. Because anti-cheat-software usually is making an exception for the big game hooking softwares out there. Like Afterburner.

**Comment by Wedmer** [comment]
Maybe just use [this] (https://obsproject.com/)?

**Comment by Jaex** [comment]
@master131
http://blogs.msdn.com/b/dsui_team/archive/2013/03/25/ways-to-capture-the-screen.aspx
(1) there tells Desktop Duplication API only can capture from Directx 11.1 > games.

**Comment by bladecoding** [comment]
Have you looked into using the hooks OBS does for game capture? If not then I may try implementing this using OBS hooks. Granted I have experienced OBS crashing a game(fallout 3). Though I thinking supporting most games is good enough(at least for me).

**Comment by Jaex** [comment]
@bladecoding you were in my Steam friends list before but can't find you now, did you removed me?

**Comment by bladecoding** [comment]
@Jaex Yes, sorry, my friends list was getting unruly with over 300+ people on it so I just removed everyone. Sent a friend request.

**Comment by Vastrix** [comment]
> It not gonna happen this is extremely difficult thing to make.

![](https://a.dilcdn.com/bl/wp-content/uploads/sites/25/2015/09/01-kuzco-crying.gif)
Shame ;(

**Comment by classyjakey** [comment]
Stupid idea 1: Make program borderless, then screenshot, then fullscreen it again?

**Comment by zeroxxx2** [comment]
This is doable.

http://hearthstonetracker.com/

Check this github project as well. Ask the author to let you take the module for hooking.

**Comment by Jaex** [comment]
That project using: https://github.com/spazzarama/Direct3DHook

**Comment by DoumanAsh** [comment]
Well, i found some hints on OpenGL games.
http://stackoverflow.com/questions/5844858/how-to-take-screenshot-in-opengl

It seems that it is possible to get pixel data using [glReadPixels](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glReadPixels.xhtml)

For DirectX i haven't found any hope...

**Comment by bladecoding** [comment]
@DoumanAsh Getting a screenshot from a game is not really that difficult. You can easily inject the OBS DLLs and grab the frames from memory. The problems are when do you hook the game, do you capture every frame for a responsive capture at the cost of performance and can you guarantee the hook won't trigger an anti-cheat.

**Comment by McoreD** [comment]
Just use Win+G in Windows 10 and have a watch folder to the save path?

**Comment by DoumanAsh** [comment]
@bladecoding The problem with anti-cheat is difficult one... Valve wouldn't let you know details of their algorithms, but i think it might possible to get in contact with them and try to find out...
But then again it seems rather overly complex to implement and may not worth it

**Comment by zeroxxx2** [comment]
Do not worry about VAC. They have their own detections, worrying about it
will not go anywhere. I have never heard people getting banned for putting
an overlay, let alone screenshot or recording programs.

**Comment by staplerx300** [comment]
People are able to use newer NVIDIA graphics cards to make WebMs while playing full screened games, where you can go into the past like 10 seconds or 10 minutes or whatever interval to splice out a recording of some kind.

If ShareX isn't planning on doing the OBS fork thing, does anyone know of a similar way to do fullscreen recording into video? I don't really want to buy a new graphics card just for that lol.

**Comment by Wedmer** [comment]
Use OBS. Also Raptr has built in Plays client and recorder.

**Comment by aglasson** [comment]
Perhaps an integration component with existing game capture technologies such as nVidia and OBS etc. Allow monitoring of their output files to use with ShareX's uploading functionality.
